homeassistant/components/lock/template.py

```diff
@@ -0,0 +1,147 @@
+    vol.Required(CONF_VALUE_TEMPLATE): cv.template,
+    vol.Optional(CONF_OPTIMISTIC, default=DEFAULT_OPTIMISTIC): cv.boolean
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices,
+                         discovery_info=None):
+    """Set up the Template lock."""
+    name = config.get(CONF_NAME)
+    value_template = config.get(CONF_VALUE_TEMPLATE)
+    value_template.hass = hass
+    value_template_entity_ids = value_template.extract_entities()
+
+    if value_template_entity_ids == MATCH_ALL:
+        _LOGGER.warning(
+            'Template lock %s has no entity ids configured to track nor '
+            'were we able to extract the entities to track from the %s '
+            'template. This entity will only be able to be updated '
+            'manually.', name, CONF_VALUE_TEMPLATE)
+
+    async_add_devices([TemplateLock(
+        hass,
+        name,
+        value_template,
+        value_template_entity_ids,
+        config.get(CONF_LOCK),
+        config.get(CONF_UNLOCK),
+        config.get(CONF_OPTIMISTIC)
+    )])
+
+
+class TemplateLock(LockDevice):
+    """Representation of a template lock."""
+
+    def __init__(self, hass, name, value_template, entity_ids,
+                 command_lock, command_unlock, optimistic):
+        """Initialize the lock."""
+        self._state = None
+        self._hass = hass
+        self._name = name
+        self._state_template = value_template
+        self._state_entities = entity_ids
+        self._command_lock = Script(hass, command_lock)
+        self._command_unlock = Script(hass, command_unlock)
+        self._optimistic = optimistic
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def template_lock_state_listener(entity, old_state, new_state):
+            """Handle target device state changes."""
+            self.async_schedule_update_ha_state(True)
+
+        @callback
+        def template_lock_startup(event):
+            """Update template on startup."""
+            if self._state_entities != MATCH_ALL:
+                # Track state change only for valid templates
+                async_track_state_change(
+                    self._hass, self._state_entities,
+                    template_lock_state_listener)
+            self.async_schedule_update_ha_state(True)
+
+        self._hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, template_lock_startup)
+
+    @property
+    def assumed_state(self):
+        """Return true if we do optimistic updates."""
+        return self._optimistic
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the lock."""
+        return self._name
+
+    @property
+    def is_locked(self):
+        """Return true if lock is locked."""
+        return self._state
+
+    @asyncio.coroutine
```

Comment:
We've migrated to `async`/`await` syntax. Please drop this decorator and prefix async methods with `async` like `async def async_update `. Replace any `yield from` with `await`.